File: app/src/main/java/com/matchthree/game/engine/GameEngine.kt
package com.matchthree.game.engine

import com.matchthree.game.model.Board
import com.matchthree.game.model.BoardConfig
import com.matchthree.game.model.Gem
import com.matchthree.game.model.GemType
import com.matchthree.game.model.Position
import com.matchthree.game.model.Special
import com.matchthree.game.rng.SeededRandom

/**
 * The pure-Kotlin game engine (AGENTS.md: engine is pure Kotlin, no Android
 * imports under game/). Consumes a board + swap, and emits an ordered list of
 * [Step]s which the UI plays back for animation.
 *
 * M4: the resolution loop also handles special gems — births (one gem per
 * shape group transforms, precedence 5 > T/L > 4 > 3), cascade-swept
 * detonation, player swap combos, and Hypercube+Hypercube board regeneration.
 */
class GameEngine(
    private val config: BoardConfig = BoardConfig(),
    private val rng: SeededRandom,
) {
    private val idSource = IdSource()

    /** A fresh board satisfying generation invariants (no match, >=1 legal move). */
    fun newGame(): Board = BoardGenerator(
        width = config.width,
        height = config.height,
        gemTypeCount = config.gemTypeCount,
        rng = rng,
        idSource = idSource,
    ).newBoard()

    /**
     * True if swapping these adjacent cells would create a match, OR the swap
     * puts two specials into contact, OR a Hypercube contacts any gem
     * (MECHANICS.md: hypercube trigger and the combo table).
     */
    fun isLegalSwap(board: Board, a: Position, b: Position): Boolean {
        if (!board.isInside(a) || !board.isInside(b)) return false
        if (!a.isOrthogonallyAdjacentTo(b)) return false
        if (board.gemAt(a) == null || board.gemAt(b) == null) return false
        val swapped = board.withSwapped(a, b)
        if (MatchDetector.findMatches(swapped).isNotEmpty()) return true
        return SpecialRules.swapContactLegal(swapped.gemAt(a), swapped.gemAt(b))
    }

    /**
     * Resolves a legal swap into the full cascade of steps, ending with a stable
     * board that contains no matches. Returns null when the swap is illegal —
     * the caller decides how to animate the rejection.
     */
    fun resolveSwap(board: Board, a: Position, b: Position): Resolution? {
        if (!isLegalSwap(board, a, b)) return null

        val steps = mutableListOf<Step>()
        steps += Step.Swap(a, b)

        var current = board.withSwapped(a, b)
        var cascadeDepth = 0
        var rounds = 0

        // 1) Player-swap special activation: combos and hypercube triggers fire
        //    before any match detection (depth-1 round).
        val swapActivation = swapActivation(current, a, b)
        if (swapActivation != null) {
            cascadeDepth = 1
            steps += Step.ComboActivate(
                specialA = swapActivation.specialA,
                specialB = swapActivation.specialB,
                affectedCells = swapActivation.affected,
            )
            steps += Step.Destroy(swapActivation.affected)
            steps += Step.Score(
                Scorer.roundScore(swapActivation.affected, cascadeDepth),
                cascadeDepth,
            )

            // Hypercube+Hypercube: full-board clear is followed by an immediate
            // regeneration (invariant-checked by the generator), not a refill.
            if (swapActivation.regenerate) {
                val fresh = newGame()
                steps += Step.Settled(fresh)
                return Resolution(board = fresh, steps = steps.toList())
            }

            val gravity = Gravity.apply(current, swapActivation.affected)
            if (gravity.falls.isNotEmpty()) steps += Step.Fall(gravity.falls)

            val refill = Refill.fill(
                board = gravity.board,
                gemTypeCount = config.gemTypeCount,
                nextId = idSource::next,
                gemType = { count -> GemType.fromIndex(rng.nextInt(count)) },
            )
            if (refill.spawned.isNotEmpty()) steps += Step.Spawn(refill.spawned)
            current = refill.board
            rounds++
        }

        // 2) Standard cascade loop: match -> (births, swept detonations) ->
        //    gravity -> refill -> re-check, until the board is stable.
        while (true) {
            val matches = MatchDetector.findMatches(current)
            if (matches.isEmpty()) break
            cascadeDepth++

            val matched = matches.flatMap { it.positions }.toSet()
            val extra = SpecialRules.sweptBlastCells(current, matched)
            val destroyed = matched + extra

            val births = SpecialRules.resolveBirths(current, matches)
            val destroyedEx = destroyed - births.map { it.cell }.toSet()

            steps += Step.Destroy(destroyedEx)
            steps += Step.Score(Scorer.roundScore(destroyedEx, cascadeDepth), cascadeDepth)

            val gravity = Gravity.apply(current, destroyedEx)
            if (gravity.falls.isNotEmpty()) steps += Step.Fall(gravity.falls)

            // Birth gems are surviving gems: apply their transformations on the
            // fallen board (same ids, same colors, new special kinds). These
            // states are impossible by construction (birth cells are excluded
            // from destruction), so invariant violations abort loudly.
            var afterBirth = gravity.board
            for (birth in births) {
                val finalPos = checkNotNull(afterBirth.positionOf(birth.gemId)) {
                    "birth gem ${birth.gemId} vanished after gravity"
                }
                val gem = checkNotNull(afterBirth.gemAt(finalPos)) {
                    "birth cell $finalPos is empty on the fallen board"
                }
                afterBirth = afterBirth.withGem(
                    finalPos,
                    Gem(gem.id, gem.type, birth.special),
                )
                steps += Step.SpecialBirth(finalPos, gem.id, birth.special)
            }

            val refill = Refill.fill(
                board = afterBirth,
                gemTypeCount = config.gemTypeCount,
                nextId = idSource::next,
                gemType = { count -> GemType.fromIndex(rng.nextInt(count)) },
            )
            if (refill.spawned.isNotEmpty()) steps += Step.Spawn(refill.spawned)

            current = refill.board
            rounds++
            if (rounds > MAX_CASCADE_ROUNDS) {
                error("cascade did not settle after $MAX_CASCADE_ROUNDS rounds")
            }
        }

        steps += Step.Settled(current)
        return Resolution(board = current, steps = steps.toList())
    }

    /**
     * Fisher-Yates reshuffle of the existing gem multiset (MECHANICS.md):
     * re-validates that the result has no pre-existing match and at least one
     * legal move. Retries up to [MAX_RESHUFFLE_ATTEMPTS]; returns null when the
     * board stays dead, which triggers game over in Zen mode.
     */
    fun reshuffle(board: Board): Board? {
        repeat(MAX_RESHUFFLE_ATTEMPTS) {
            val candidate = shuffleMultiset(board)
            if (MatchDetector.findMatches(candidate).isEmpty() && LegalMoveDetector.hasLegalMove(candidate)) {
                return candidate
            }
        }
        return null
    }

    /** Shuffles the gems into a fresh layout, preserving ids and types. */
    private fun shuffleMultiset(board: Board): Board {
        val gems = board.positions().mapNotNull { board.gemAt(it) }.toMutableList()
        for (i in gems.size - 1 downTo 1) {
            val j = rng.nextInt(i + 1)
            val tmp = gems[i]
            gems[i] = gems[j]
            gems[j] = tmp
        }
        var next = 0
        return Board.create(board.width, board.height) { position ->
            if (next < gems.size) gems[next++] else null
        }
    }

    /**
     * Computes what (if anything) a player swap of two gems activates: a combo
     * between two specials, or a single hypercube trigger against a normal gem.
     * Returns null when neither gem is a special (plain match path).
     */
    private fun swapActivation(
        board: Board,
        a: Position,
        b: Position,
    ): SwapActivation? {
        val gemA = board.gemAt(a)
        val gemB = board.gemAt(b)
        val specA = gemA?.special
        val specB = gemB?.special
        if (specA == null && specB == null) return null

        // Hypercube against a plain gem: clears every gem of the swapped color
        // plus the hypercube itself (single-activator, no combo partner).
        if (specA == Special.HYPERCUBE && specB == null || specB == Special.HYPERCUBE && specA == null) {
            val hyperPos = if (specA == Special.HYPERCUBE) a else b
            val otherPos = if (hyperPos == a) b else a
            val partner = board.gemAt(otherPos)
            if (partner == null || partner.special != null) return null
            return SwapActivation(
                specialA = Special.HYPERCUBE,
                specialB = null,
                affected = SpecialRules.hypercubeTriggerCells(board, hyperPos, otherPos, null),
                regenerate = false,
            )
        }

        // Combo table: the two swapped gems are both specials.
        if (specA != null && specB != null) {
            val affected = SpecialRules.comboAffectedCells(board, a, b, specA, specB)
            val regenerate = specA == Special.HYPERCUBE && specB == Special.HYPERCUBE
            return SwapActivation(specA, specB, affected, regenerate)
        }
        return null
    }

    /** Result of a player-swap special activation. */
    private data class SwapActivation(
        val specialA: Special,
        val specialB: Special?,
        val affected: Set<Position>,
        val regenerate: Boolean,
    )

    private companion object {
        /** Safety valve against an unlucky refill streak looping forever. */
        const val MAX_CASCADE_ROUNDS = 1_000

        /** MECHANICS.md: reshuffle retries before handing the board to game over. */
        const val MAX_RESHUFFLE_ATTEMPTS = 20
    }
}

/** Outcome of [GameEngine.resolveSwap]: the settled board plus its playback steps. */
data class Resolution(val board: Board, val steps: List<Step>)

File: Engine/Rules/GameEngine.cs
using MatchThree.Engine.Model;

namespace MatchThree.Engine.Rules;

public sealed class GameEngine
{
    private const int MAX_CASCADE_ROUNDS = 1_000;
    private const int MAX_RESHUFFLE_ATTEMPTS = 20;

    private readonly BoardConfig config;
    private readonly SeededRandom rng;
    private readonly IdSource idSource = new();

    public GameEngine(BoardConfig config, SeededRandom rng)
    {
        this.config = config;
        this.rng = rng;
    }

    public GameEngine(SeededRandom rng) : this(new BoardConfig(), rng)
    {
    }

    public Board NewGame() =>
        new BoardGenerator(
            width: config.Width,
            height: config.Height,
            gemTypeCount: config.GemTypeCount,
            rng: rng,
            idSource: idSource
        ).NewBoard();

    public bool IsLegalSwap(Board board, Position a, Position b)
    {
        if (!board.IsInside(a) || !board.IsInside(b)) return false;
        if (!a.IsOrthogonallyAdjacentTo(b)) return false;
        if (board.GemAt(a) is null || board.GemAt(b) is null) return false;

        var swapped = board.WithSwapped(a, b);
        if (MatchDetector.FindMatches(swapped).Count > 0) return true;
        return SpecialRules.SwapContactLegal(swapped.GemAt(a), swapped.GemAt(b));
    }

    public Resolution? ResolveSwap(Board board, Position a, Position b)
    {
        if (!IsLegalSwap(board, a, b)) return null;

        var steps = new List<Step>();
        steps.Add(new Step.Swap(a, b));

        var current = board.WithSwapped(a, b);
        var cascadeDepth = 0;
        var rounds = 0;

        var swapActivation = ComputeSwapActivation(current, a, b);
        if (swapActivation is not null)
        {
            cascadeDepth = 1;
            steps.Add(new Step.ComboActivate(
                SpecialA: swapActivation.SpecialA,
                SpecialB: swapActivation.SpecialB,
                AffectedCells: swapActivation.Affected));
            steps.Add(new Step.Destroy(swapActivation.Affected));
            steps.Add(new Step.Score(
                Scorer.RoundScore(swapActivation.Affected, cascadeDepth),
                cascadeDepth));

            if (swapActivation.Regenerate)
            {
                var fresh = NewGame();
                steps.Add(new Step.Settled(fresh));
                return new Resolution(Board: fresh, Steps: steps);
            }

            var gravity = Gravity.Apply(current, swapActivation.Affected);
            if (gravity.Falls.Count > 0) steps.Add(new Step.Fall(gravity.Falls));

            var refill = Refill.Fill(
                board: gravity.Board,
                gemTypeCount: config.GemTypeCount,
                nextId: idSource.Next,
                gemType: count => GemTypes.FromIndex(rng.NextInt(count)));
            if (refill.Spawned.Count > 0) steps.Add(new Step.Spawn(refill.Spawned));
            current = refill.Board;
            rounds++;
        }

        while (true)
        {
            var matches = MatchDetector.FindMatches(current);
            if (matches.Count == 0) break;
            cascadeDepth++;

            var matched = matches.SelectMany(m => m.Positions).ToHashSet();
            var extra = SpecialRules.SweptBlastCells(current, matched);
            var destroyed = new HashSet<Position>(matched.Concat(extra));

            var births = SpecialRules.ResolveBirths(current, matches);
            var birthCells = births.Select(b => b.Cell).ToHashSet();
            var destroyedEx = new HashSet<Position>(destroyed.Except(birthCells));

            steps.Add(new Step.Destroy(destroyedEx));
            steps.Add(new Step.Score(Scorer.RoundScore(destroyedEx, cascadeDepth), cascadeDepth));

            var gravity = Gravity.Apply(current, destroyedEx);
            if (gravity.Falls.Count > 0) steps.Add(new Step.Fall(gravity.Falls));

            var afterBirth = gravity.Board;
            foreach (var birth in births)
            {
                var finalPos = afterBirth.PositionOf(birth.GemId)
                    ?? throw new InvalidOperationException($"birth gem {birth.GemId} vanished after gravity");
                var gem = afterBirth.GemAt(finalPos)
                    ?? throw new InvalidOperationException($"birth cell {finalPos} is empty on the fallen board");
                afterBirth = afterBirth.WithGem(finalPos, new Gem(gem.Id, gem.Type, birth.Special));
                steps.Add(new Step.SpecialBirth(finalPos, gem.Id, birth.Special));
            }

            var refill = Refill.Fill(
                board: afterBirth,
                gemTypeCount: config.GemTypeCount,
                nextId: idSource.Next,
                gemType: count => GemTypes.FromIndex(rng.NextInt(count)));
            if (refill.Spawned.Count > 0) steps.Add(new Step.Spawn(refill.Spawned));

            current = refill.Board;
            rounds++;
            if (rounds > MAX_CASCADE_ROUNDS)
            {
                throw new InvalidOperationException($"cascade did not settle after {MAX_CASCADE_ROUNDS} rounds");
            }
        }

        steps.Add(new Step.Settled(current));
        return new Resolution(Board: current, Steps: steps);
    }

    public Board? Reshuffle(Board board)
    {
        for (var attempt = 0; attempt < MAX_RESHUFFLE_ATTEMPTS; attempt++)
        {
            var candidate = ShuffleMultiset(board);
            if (MatchDetector.FindMatches(candidate).Count == 0 && LegalMoveDetector.HasLegalMove(candidate))
            {
                return candidate;
            }
        }
        return null;
    }

    private Board ShuffleMultiset(Board board)
    {
        var gems = board.Positions().Select(p => board.GemAt(p)).Where(g => g is not null).Select(g => g!.Value).ToList();
        for (var i = gems.Count - 1; i >= 1; i--)
        {
            var j = rng.NextInt(i + 1);
            (gems[i], gems[j]) = (gems[j], gems[i]);
        }
        var next = 0;
        return Board.Create(board.Width, board.Height, _ =>
            next < gems.Count ? gems[next++] : null);
    }

    private SwapActivation? ComputeSwapActivation(Board board, Position a, Position b)
    {
        var gemA = board.GemAt(a);
        var gemB = board.GemAt(b);
        var specA = gemA?.Special;
        var specB = gemB?.Special;
        if (specA is null && specB is null) return null;

        if ((specA == Special.Hypercube && specB is null) ||
            (specB == Special.Hypercube && specA is null))
        {
            var hyperPos = specA == Special.Hypercube ? a : b;
            var otherPos = hyperPos == a ? b : a;
            var partner = board.GemAt(otherPos);
            if (partner is null || partner.Value.Special is not null) return null;
            return new SwapActivation(
                SpecialA: Special.Hypercube,
                SpecialB: null,
                Affected: SpecialRules.HypercubeTriggerCells(board, hyperPos, otherPos, null),
                Regenerate: false);
        }

        if (specA is not null && specB is not null)
        {
            var affected = SpecialRules.ComboAffectedCells(board, a, b, specA.Value, specB.Value);
            var regenerate = specA == Special.Hypercube && specB == Special.Hypercube;
            return new SwapActivation(specA.Value, specB.Value, affected, regenerate);
        }
        return null;
    }

    private sealed record SwapActivation(
        Special SpecialA,
        Special? SpecialB,
        ISet<Position> Affected,
        bool Regenerate);
}

public sealed record Resolution(Board Board, List<Step> Steps);